feat(oauth2) add ability to persist refresh tokens

amberheilman opened the pull request:
### Summary
Add support for persisting oauth2 refresh tokens. Follows guidelines in [RFC6749](https://tools.ietf.org/html/rfc6749#section-6) to allow for the optional persistence of the refresh token when refreshing an access token.

### Related discussion
https://discuss.konghq.com/t/long-lived-refresh-tokens/659
https://discuss.konghq.com/t/long-lived-oauth-2-refresh-tokens/3005

Commit:
feat(oauth2) add ability to persist refresh tokens

Adds a configuration option that will persist the
refresh token value throughout the lifecycle of
the token. If `persistent_refresh_token` is
true, the refresh token generated with the first
access token will be used forever. RFC6749 states
that the refresh token MAY be regenerated, but
it is not required. In the case of a refresh, if
a client fails to receive the new access token
and a newly generated refresh token (maybe due to
a timeout) the client is left without the ability
to recover and must re-authenticate. When using a
persistent refresh token, the refresh can be
re-attempted.

https://tools.ietf.org/html/rfc6749#section-6
https://discuss.konghq.com/t/long-lived-refresh-tokens/659
https://discuss.konghq.com/t/long-lived-oauth-2-refresh-tokens/3005

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/kong?pullRequest=5264) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/kong?pullRequest=5264) before we can accept your contribution.<br/><hr/>**amberh** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/kong?pullRequest=5264) it.</sub>

bob983 commented:
This is a nice coincidence, we started chatting about similar issue here https://github.com/Kong/kong/issues/5252.

dave-shawley commented:
@bob983 - we ran into the disappearing token problem using another company's OAuth 2 API and decided that we simply couldn't put our customers through the same pain.  Hopefully this will meet your needs as well.  We briefly looked into supporting multiple linked tokens side-by-side instead of making them persistent and it was quite a bit more work than expected.  Ultimately, making refresh tokens immutable until revocation met our needs.

We did implement [RFC-7009 token revocation] as a separate service using the `DELETE /oauth2_tokens/:token_id` Admin API endpoint.  That handled the need to explicitly revoke either an access token or a refresh token mentioned in #5652 .  We could look into contributing that back if there is interest.

[RFC-7009 token revocation]: https://tools.ietf.org/html/rfc7009

bob983 commented:
@dave-shawley yeah, this would definitely work for us well :) 

We handle the token revocation via our own UI which in turn calls the Kong Admin API and deletes the refresh token / acess token pair and thus RFC-7009 is not necessary for us but it would be a nice addition.

bob983 commented:
Any news on this PR? I'm eagerly watching it :)

bungle reviewed (changes requested):
This is bit off-topic, but mind you change these as well, perhaps another commit:
refactor(oauth2) change generate_token parameters

bungle commented:
In general I think this is good... only thing I am hesitating is the quite long and perhaps a bit misleading (as all the refresh tokens are persistent, but not reusable currently):
```
persistent_refresh_token
```

We already have:
```
config.refresh_token_ttl
```

@dave-shawley, what do you think e.g.:_
```
config.refresh_token_reuse=true
```
or
```
config.refresh_token_reusable=true
```
or
```
config.reuse_refresh_token=true
```

Pick what you like the best. I think after that we can go ahead and merge. Please also rebase to latest `next` and change the PR to be merged to `next` as that is where `feats` go.

Thank you for the PR. I have to say that these are very high in quality. We should have reviewed this earlier. I'll focus to get these in!

Commit:
refactor(oauth2) change generate_token parameters

The expiration parameter was removed since it was not
used in any calling code.

feat(oauth2) add ability to persist refresh tokens #5264

Commit:
refactor(oauth2) rename config option

Rename the `persistent_refresh_token` configuration option
to `reuse_refresh_token` for better clarity since all refresh
tokens are persisted.

feat(oauth2) add ability to persist refresh tokens #5264

amberheilman commented:
@bungle,
PR has been rebased against next. 
Changed the configuration option from `persistent_refresh_token` to `reuse_refresh_token` as well.

bungle commented:
@amberheilman, thank you sir! it is finally merged now, and will land on `2.1`.
